### PR TITLE
fix(core): key navigation in vertical navigation

### PR DESCRIPTION
**Pull request by shrvr**
## Related Issue(s)

Part of #8557 

## Description

1. Added arrow key navigation. (Not for condensed mode)
2. Added list expansion using Arrow, Space & Enter key.

### CODE REVIEW

**Review by N1XUS — APPROVED** [review]

**Comment by N1XUS** [comment]
Good job! One question: if we have keyboard navigation, and by pressing enter on focused item expand the subtree, do we actually need to have focusable button?
image

**Comment by shrvr** [comment]
> 



> Good job! One question: if we have keyboard navigation, and by pressing enter on focused item expand the subtree, do we actually need to have focusable button? image

@N1XUS As the use of the button is in user's hand, I think user can make it tabbable/un-tabbable.

File: libs/core/src/lib/list/list-navigation-item/list-navigation-item.component.ts [hunk 1]
```diff
@@ -57,6 +70,19 @@ export class ListNavigationItemComponent implements AfterContentInit {
     /** @hidden */
     _innerText: string;
 
+    /** @hidden
+     * false if list-item is within unexpanded list (not visible to user until list expanded). default is true
+     */
+    _isItemVisible = true;
+
+    /** @hidden */
+    readonly _focused$ = new Subject<{ focusedWithin: boolean }>();
```
Comment (platon-rov) [inline libs/core/src/lib/list/list-navigation-item/list-navigation-item.component.ts:79]: can be simple `<boolean>`.

File: libs/core/src/lib/list/list-navigation-item/list-navigation-item.component.ts [hunk 2]
```diff
@@ -89,11 +157,21 @@ export class ListNavigationItemComponent implements AfterContentInit {
         return retVal;
     }
 
+    focus(): void {
```
Comment (platon-rov) [inline libs/core/src/lib/list/list-navigation-item/list-navigation-item.component.ts:160]: typedocs?

File: libs/core/src/lib/vertical-navigation/vertical-navigation.component.ts [hunk 3]
```diff
@@ -21,16 +24,86 @@ export class VerticalNavigationComponent implements AfterContentInit {
     @Input()
     condensed = false;
 
+    /** An RxJS Subject that will kill the data stream upon component’s destruction (for unsubscribing)  */
+    private readonly _onDestroy$: Subject<void> = new Subject<void>();
```
Comment (platon-rov) [inline libs/core/src/lib/vertical-navigation/vertical-navigation.component.ts:28]: minor: `private readonly _onDestroy$ = new Subject<void>()`, type will be inferred.

File: libs/core/src/lib/vertical-navigation/vertical-navigation.component.ts [hunk 4]
```diff
@@ -21,16 +24,86 @@ export class VerticalNavigationComponent implements AfterContentInit {
...
     /** @hidden */
-    @ContentChildren(ListNavigationItemComponent)
+    @ContentChildren(ListNavigationItemComponent, { descendants: true })
     private _navigationItems: QueryList<ListNavigationItemComponent>;
 
+    /** @hidden
+     * Querylist of list-items in main navigation.
+     */
+    @ContentChildren(ListNavigationItemComponent)
+    private _mainNavigationItems: QueryList<ListNavigationItemComponent>;
```
Comment (platon-rov) [inline libs/core/src/lib/vertical-navigation/vertical-navigation.component.ts:41]: `_directNavigationItems`?

File: libs/core/src/lib/vertical-navigation/vertical-navigation.component.ts [hunk 5]
```diff
@@ -21,16 +24,86 @@ export class VerticalNavigationComponent implements AfterContentInit {
...
+        } else {
+            this._keyManager.setActiveItem(index);
+        }
+    }
+
+    /** @hidden */
+    private _listenOnQueryChange(): void {
+        this._navigationItems.changes.pipe(startWith(0), takeUntil(this._onDestroy$)).subscribe(() => {
+            this._listenOnItemsClick();
+            setTimeout(() => {
```
Comment (platon-rov) [inline libs/core/src/lib/vertical-navigation/vertical-navigation.component.ts:73]: minor: `delay(0)` in pipe would look a little bit cleaner.

File: libs/core/src/lib/vertical-navigation/vertical-navigation.component.ts [hunk 6]
```diff
@@ -21,16 +24,86 @@ export class VerticalNavigationComponent implements AfterContentInit {
...
+    }
+
+    /** @hidden */
+    private _listenOnItemsClick(): void {
+        /** Finish all the streams, from before */
+        this._onRefresh$.next();
+
+        /** Merge refresh/destroy observables */
+        const completion$ = merge(this._onRefresh$, this._onDestroy$);
+        const interactionChangesIndexes: Observable<{ index: number; updateOnly: boolean }>[] =
```
Comment (platon-rov) [inline libs/core/src/lib/vertical-navigation/vertical-navigation.component.ts:89]: minor: probably type will be inferred.

**Review by platon-rov — APPROVED** [review]
left minor items, good job

File: libs/core/src/lib/list/list-navigation-item/list-navigation-item.component.ts [hunk 7]
```diff
@@ -89,11 +157,21 @@ export class ListNavigationItemComponent implements AfterContentInit {
...
+        this._elementRef.nativeElement.focus();
+    }
+
+    click(): void {
```
Comment (InnaAtanasova) [inline libs/core/src/lib/list/list-navigation-item/list-navigation-item.component.ts:164]: here as well

**Review by InnaAtanasova — COMMENTED** [review]
Please address the comments from Platon, the rest looks good

File: libs/core/src/lib/vertical-navigation/vertical-navigation.component.ts (hunk 4 above)
Comment (shrvr) [reply libs/core/src/lib/vertical-navigation/vertical-navigation.component.ts:41]: I am referring to immediate list-items in main navigation. @platon-rov 

image

File: libs/core/src/lib/list/list-navigation-item/list-navigation-item.component.ts [hunk 8]
```diff
@@ -70,13 +97,53 @@ export class ListNavigationItemComponent implements AfterContentInit {
...
     @HostListener('click', ['$event'])
     onItemClick(event: MouseEvent): void {
         event.stopPropagation();
         this._handleExpandedChanges(!this.expanded);
     }
 
+    /** @hidden */
+    @HostListener('keydown', ['$event'])
+    keyDownHandler(event: KeyboardEvent): void {
+        if (KeyUtil.isKeyCode(event, [RIGHT_ARROW])) {
```
Comment (mikerodonnell89) [inline libs/core/src/lib/list/list-navigation-item/list-navigation-item.component.ts:115]: Really nice work! Only thing I would change is I think these left/right arrow checks need to check for RTL mode, I think a left arrow press should expand and a right arrow should collapse when in RTL mode. Other than that I think ready to merge

**Review by mikerodonnell89 — CHANGES_REQUESTED** [review]

File: libs/core/src/lib/list/list-navigation-item/list-navigation-item.component.ts (hunk 8 above)
Comment (shrvr) [reply libs/core/src/lib/list/list-navigation-item/list-navigation-item.component.ts:115]: Hm, agree. It should also be enabled for [Side Navigation](https://fundamental-ngx.netlify.app/#/core/side-navigation#programmatic-state-change).

**Review by mikerodonnell89 — APPROVED** [review]
Not working in condensed mode, but let's consider that a separate issue https://github.com/SAP/fundamental-ngx/issues/8761

**Pull request merged**